### PR TITLE
Add the capability to disallow loops in a adaptive auth script

### PR DESCRIPTION
**Pull request by janakamarasena**
### Proposed changes in this pull request

This PR provides  the capability to disallow loops in an adaptive auth script. 
If loops are not allowed and an authentication script is added with a loop an error message will be shown.

Add the below config to deployment.toml to disallow loops
```
[authentication.adaptive]
allow_loops = false
```
Default value: true (loops are allowed)

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java [hunk 1]
```diff
@@ -473,4 +486,54 @@ private boolean isGroup(String localRoleName) {
...
+        }
+
+        if (!allowLoopsInAdaptiveAuthScript()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Loops are not allowed in the authentication script. " +
+                        "Therefore checking whether loops are present in the provided script.");
+            }
+            Matcher matcher = loopPattern.matcher(script);
+            if (matcher.find()) {
+                validationMsg.add("Loops are not allowed in the adaptive authentication script.");
```
Comment (madurangasiriwardena) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java:508]:
```suggestion
                validationMsg.add("Loops are not allowed in the adaptive authentication script, but loop is available in the provided script.");
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java [hunk 2]
```diff
@@ -473,4 +486,54 @@ private boolean isGroup(String localRoleName) {
...
+            }
+        }
+    }
+
+    private boolean allowLoopsInAdaptiveAuthScript() {
```
Comment (senthalan) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java:513]:
```suggestion
    private boolean isLoopsInAdaptiveAuthScriptAllowed() {
```
?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java [hunk 3]
```diff
@@ -473,4 +486,54 @@ private boolean isGroup(String localRoleName) {
         return !Stream.of(INTERNAL_DOMAIN, APPLICATION_DOMAIN, WORKFLOW_DOMAIN).anyMatch(domain -> localRoleName
                 .toUpperCase().startsWith((domain + UserCoreConstants.DOMAIN_SEPARATOR).toUpperCase()));
     }
+
+    private void validateAdaptiveAuthScript(List<String> validationMsg,
```
Comment (madurangasiriwardena) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java:490]:
```suggestion
    private void validateAdaptiveAuthScript(List<String> validationErrors,
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java (hunk 2 above)
Comment (janakamarasena) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java]: updated

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java (hunk 1 above)
Comment (janakamarasena) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java]: updated

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java (hunk 3 above)
Comment (janakamarasena) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/validator/DefaultApplicationValidator.java]: updated

**Review by senthalan — APPROVED** [review]

**Pull request merged**